packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-tree.ts

```diff
@@ -0,0 +1,109 @@
+import { BulkEditNodeSelection } from './bulk-edit-node-selection';
+import URI from '@theia/core/lib/common/uri';
+
+@injectable()
+export class BulkEditTree extends TreeImpl {
+    public async initTree(workspaceEdit: monaco.languages.WorkspaceEdit, fileContents: Map<string, string>): Promise<void> {
+        this.root = <CompositeTreeNode>{
+            visible: false,
+            id: 'theia-bulk-edit-tree-widget',
+            name: 'BulkEditTree',
+            children: this.getChildren(workspaceEdit, fileContents),
+            parent: undefined
+        };
+    }
+
+    private getChildren(workspaceEdit: monaco.languages.WorkspaceEdit, fileContentsMap: Map<string, string>): BulkEditInfoNode[] {
+        let bulkEditInfos: BulkEditInfoNode[] = [];
+        if (workspaceEdit.edits) {
+            bulkEditInfos = workspaceEdit.edits
+                .map(edit => this.getResourcePath(edit))
+                .filter((path, index, arr) => path && arr.indexOf(path) === index)
+                .map((path: string) => this.createBulkEditInfo(path, new URI(path), fileContentsMap.get(path)))
+                .filter(Boolean);
+
+            if (bulkEditInfos.length > 0) {
+                bulkEditInfos.forEach(editInfo => {
+                    editInfo.children = workspaceEdit.edits.filter(edit => ((('resource' in edit) && edit.resource.path === editInfo.id)) ||
+                        (('newUri' in edit) && edit.newUri && edit.newUri.path === editInfo.id))
+                        .map((edit, index) => this.createBulkEditNode(edit, index, editInfo));
+                });
+            }
+        }
+        return bulkEditInfos;
+    }
+
+    private createBulkEditNode(bulkEdit: monaco.languages.WorkspaceTextEdit | monaco.languages.WorkspaceFileEdit, index: number, parent: BulkEditInfoNode): BulkEditNode {
+        const id = parent.id + '_' + index;
+        const existing = this.getNode(id);
+        if (BulkEditNode.is(existing)) {
+            existing.bulkEdit = bulkEdit;
+            return existing;
+        }
+        return {
+            id,
+            name: 'bulkEdit',
+            parent,
+            selected: false,
+            uri: parent.uri,
+            bulkEdit
+        };
+    }
+
+    private createBulkEditInfo(id: string, uri: URI, fileContents: string | undefined): BulkEditInfoNode {
+        return {
+            id,
+            uri,
+            expanded: true,
+            selected: false,
+            parent: this.root as BulkEditInfoNode,
+            fileContents,
+            children: []
+        };
+    }
+
+    private getResourcePath(edit: monaco.languages.WorkspaceTextEdit | monaco.languages.WorkspaceFileEdit): string | undefined {
+        return (edit && ('resource' in edit) && edit.resource && edit.resource.path) ?
+            edit.resource.path :
+            ((edit && ('newUri' in edit) && edit.newUri && edit.newUri.path) ? edit.newUri.path : undefined);
+    }
+}
```

Comment:
I do not believe `edit` can be `undefined`, there's no need to check for it with this method signature.
We also have the following two methods which can be used:

https://github.com/eclipse-theia/theia/blob/cd85908d3bb5af06477ead7168110f2306a42145/packages/monaco/src/browser/monaco-workspace.ts#L33-L49

Reply:
Fixed